### PR TITLE
cassandra-driver: Promise interface for asynchronous Client methods

### PR DESCRIPTION
**Pull request by zenlor**
…hods

Please fill in this template.

- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `tsc` without errors.
- [x] Run `npm run lint package-name` if a `tslint.json` is present.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: http://docs.datastax.com/en/developer/nodejs-driver/3.2/api/class.Client/
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "../tslint.json" }`.

### CODE REVIEW

**Comment by aliem** [comment]
Sadly the public interface of methods is abusing javascript's optional parameters. It may be fixed with:

```
execute(query: string, params?: any, options?: QueryOptions, callback?: ResultCallback): void|Promise<types.ResultSet>;
```

But it throws static analysis and type safety out of the window.
For *async* functions, should I simply test it using `then`?

**Comment by subk** [comment]
Any chance to get this maintained by @datastax (cc @jorgebay) ?

File: cassandra-driver/index.d.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Type definitions for nodejs-driver v0.8.2
+// Type definitions for cassandra-driver v3.0.2
```
Comment (jorgebay) [inline cassandra-driver/index.d.ts]: Do you mean v3.2.0?

**Comment by jorgebay** [comment]
I have no experience with TypeScript... we can ping here / post an issue when there is an API change.
We don't plan to make big API changes in the near future, at least not as big as what we did for `Promise` support.

Maybe we can create a script that translates jsdoc -> typescript definitions? jsdoc bin has good support for transformations.

File: cassandra-driver/index.d.ts (hunk 1 above)
Comment (aliem) [inline cassandra-driver/index.d.ts]: oh, yes, sorry for this

File: cassandra-driver/index.d.ts [hunk 2]
```diff
@@ -506,11 +506,22 @@ export interface Client extends events.EventEmitter {
...
   eachRow(query: string, params?: any, options?: QueryOptions, rowCallback?: Callback, callback?: Callback): void;
-  execute(query: string, params?: any, options?: QueryOptions, callback?: ResultCallback): void;
+  execute(query: string, params: any, options: QueryOptions, callback: ResultCallback): void;
+  execute(query: string, options: QueryOptions, callback: ResultCallback): void;
+  execute(query: string, params: any, callback: ResultCallback): void;
+  execute(query: string, callback: ResultCallback): void;
+  execute(query: string, params: any, options: QueryOptions): Promise<types.ResultSet>;
+  execute(query: string, options: QueryOptions): Promise<types.ResultSet>;
+  execute(query: string, params?: any): Promise<types.ResultSet>;
+  execute(query: string): Promise<types.ResultSet>;
```
Comment (minestarks) [inline cassandra-driver/index.d.ts]: You don't need quite this many overloads, the last 4 can be reduced down to 2.

```ts
execute(query: string, params: any, options: QueryOptions, callback: ResultCallback): void;
execute(query: string, options: QueryOptions, callback: ResultCallback): void;
execute(query: string, params: any, callback: ResultCallback): void;
execute(query: string, callback: ResultCallback): void;

execute(query: string, options: QueryOptions): Promise<types.ResultSet>;
execute(query: string, params?: any, options?: QueryOptions): Promise<types.ResultSet>;
```

Comment (jorgebay) [inline cassandra-driver/index.d.ts]: I think the different overloads expressed in TypeScript would be something like:
Callback-based
```javascript
execute(query: string, params: any, options: QueryOptions, callback: ResultCallback): void;
execute(query: string, params: any, callback: ResultCallback): void;
execute(query: string, callback: ResultCallback): void;
```
Promise based
```javascript
execute(query: string, params?: any, options?: QueryOptions): Promise<types.ResultSet>;
```

It's not allowed to specify the query options without specifying the parameters.

**Comment by paulvanbrenk** [comment]
Closing for now.

Please re-open when you've resolved the 2 issues (the version, and the number of overloads).

**Pull request closed without merging**